apollographql/apollo-server
# Embedded Explorer AS Landing Page Plugin / 🚨 Don't merge until https://github.com/apollographql/studio-landing-page/pull/30 is merged 🚨

[JIRA](https://apollographql.atlassian.net/browse/NEBULA-948)

**Context**

We want to offer a way of using Explorer with Apollo Server without the need to configure CORS. This PR allows folks to choose to embed an Explorer populated with their schema either from 1. introspection from the current endpoint or 2. graphRef.

**This PR**

Adds a plugin to `apollo-server` allowing folks to use a full page embedded explorer instead of the default landing page or other options like graphql playground. This plugin renders the studio landing page with a config. The config can contain a graphRef, or not. If a graph ref is passed - the studio landing page renders the registered graph's explorer, otherwise, the studio landing page always first receives the schema from apollo server. Otherwise, if this graph is a gateway or the user has specified a poll interval, makes introspection requests every `schemaPollIntervalMs` ms to the current href to grab the graphql schema and populate the explorer.

The config allows folks to decide 
1. the poll interval to introspect the schema
2. whether or not to keep local storage around
3. display options for the embedded explorer
4. default document, headers and vars for the embed on load

**How to test**

[Until the embed side is merged](https://github.com/apollographql/studio-landing-page/pull/30), you will need to switch the version to `mayaTest10`. 

Check out the code sandbox npm deploy preview below, and add the ApolloServerPluginLandingPageEmbeddedExplorer. For example: 

```
      ApolloServerPluginLandingPageEmbeddedExplorer({
        document: "test",
        variables: { test: "value" },
        headers: { test: "value" },
        includeCookies: false,
        footer: true,
        shouldEmbedExplorer: true,
        schemaPollIntervalMs: 10000,
        embedDisplayOptions: {
          showHeadersAndEnvVars: true,
          docsPanelState: "open",
          theme: "dark",
        },
        persistExplorerState: false,
      }),
```

**Open Questions**
1. [Folks can specify headers from the config, but can change them in the UI. We only know about headers from the config when we introspect the endpoint for your schema. What do we do?](https://apollograph.slack.com/archives/C032ZPPE5E1/p1650302303017429)
     **Answer**: We are going to post message the introspection request from the embed to studio-landing-page. That will be added to [this PR ](https://github.com/apollographql/studio-landing-page/pull/30) in studio-landing-page as well as a new PR in embedded-explorer to come. 
2. Do we need language of `prod` / `local` in here or is `graphRef` / no graph ref enough?
